### ISSUE TITLE
feature: improved pipeline

### ISSUE THREAD

**Issue by a0m0rajab** [unifyai/ivy#21311]
# Branches

The development pipeline right now in Ivy consists of opening the PR's to the master branch always and creating a fork for that, I think a better situation would be if there was a beta/development branch were all development goes there and we have a default branch for production, were we marge from development to production. This would make it easier to have an idea about what would break the tests and having the tests always passing for the production. 

When we use branches it would make it easier and smoother experience. we can even add protection rules to the production branch.

it would be great to add protection rules to the main/master branch as well.

# Tests 

From a developer's point of view, a framework with tests is awesome since you are sure it's trustworthy. Yet, having failing tests even if it's small makes a kind of bad impression, it would be great to only merge PR's passing the tests, we can have this made sure if we did a dev/prod branches from my point of view.

changing the tests name and explaining the mapping and CI/CD code would be awesome as well, so that we can have a readable tests instead of just tests going from 1 to 120. 

# Clean Up 

It would be great to check the stale issues and PR's to close them, this would make less stress and focus on the important and active ones. 

# GitHub Automation 

Some of the automation that's happening right now is happening from a different account which is ivy-seeds and ivy-leaves, it would be nice if we could use the automation system of Github: https://github.com/features/actions to be able to use that to write comments for people to help them with their PRs and issues.

We can display the summary of the test results in a comment and remove it when it's resolved, this would save time for the reviewers and make it meaningful for the contributors. 

https://docs.github.com/en/issues/planning-and-tracking-with-projects/automating-your-project/using-the-built-in-automations

# Conventional Commits 

Using and enforcing conventional commits would make a great way to have a clean history of the project, this will help to make any change understandable and give an idea of where to look if something got broken, plus it would help with changelog releases and automate it. 

https://www.conventionalcommits.org/en/v1.0.0/

# Conventional release and changelog:

It would be great if we could use the release option in GitHub, to have a release made based on the development we do, 
This would be helpful to share the updates and engage with developers

While doing this we could go with semantic versioning and having an update every two weeks: https://semver.org/

For the release this could be used to have shoutouts for the people who opened a PR for it. 

This an option: 
https://github.com/conventional-changelog/conventional-changelog
https://semantic-release.gitbook.io/semantic-release/
https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository
https://www.npmjs.com/package/semantic-release-conventional-commits

Tensorflow example: 
https://github.com/tensorflow/tensorflow/releases
Keras
https://github.com/keras-team/keras/releases
Calcom: 
https://github.com/calcom/cal.com/releases
Insights: 
https://github.com/open-sauced/insights/releases

# PR Templates: 

It would be great to have a PR template and checklist for people to use immediately we can have a template for frontend, formatting and other types to use. 

The template would make it easier for the reviewer to know what happened in the code from first look. 

# Engagement

The process of opening a new task right now is being like: 
- go to tracker task
- open a task from that
- add the task in comments 
-  work on PR 
Yet if someone was not working on the task, we will have a subtask stale and it will be closed, it will stay in the tracker task as well, 

from my perspective, it would be great to use GitHub features here,like assignee and remove assignee when it's stale, this will help us to know which tasks are the most struggling one rather than just check the tracker task. so the approach would be: 

instead of closing the tasks that are stale we can do the next: 
- open a task from the list, 
- add the task in comment,
- go to the task opened and assign yourself (we can enable this with GitHub actions)
- if it's stale we remove the assignee, and people can assign themselves if there are no assignee! 

This will make a function related conversations in one place and does not make a loss of the developers stories

# CI/CD 

Here since we are testing the pre-commit and linting in the CI/CD we can let the CI/CD commit the changes happened to the PR and add them, this will make it easier to have such fixes. 

Plus, since we have a lot of tests that run, we can let the tests that we need to read it's information to be written and added in a comment in the PR this will make it easier for the reviewer to take a look and for the coder to check their issues.
